pipeline/matching/generators/rare_intents.py

```diff
@@ -1,18 +1,138 @@
+        percentDict = self.percentage(inp)
+        # add all rare givers to this dictionary to later match them
+        # key is giving name and value is list of users
+        rareDict = {}
+
+        # check if giving side is rare by comparing it to the community threshold
+        for perecnt in percentDict:
+            if percentDict[perecnt] < self.max_frequency:
+                rareDict[perecnt] = percentDict.get(perecnt)
+
+        match: List[Match] = []
+
+        # key is giving name and value is list of users
+        rareGivingIntentUserMap = defaultdict(list)
+        rareSeekingIntentUserMap = defaultdict(list)
+
+        for user in inp.users:
+            for intent in user.intents:
+                if intent.code in rareDict and (intent.side == Side.SEEKING):
+                    rareSeekingIntentUserMap[intent.code].append(user)
+
+                if intent.code in rareDict and (intent.side == Side.GIVING):
+                    rareGivingIntentUserMap[intent.code].append(user)
+
+        # for user in inp.users:
+        #     for intent in user.intents:
+        #         if intent.code in rareSeekingIntentUserMap and (intent.side == Side.SEEKING):
+        #             rareSeekingIntentUserMap[intent.code].append(user)
+
+        # holds matches made
+        matches: Dict[str, Match] = {}
+        # append matches to a dictionary
+        # before doing so sort the id and place it as the key so it could be cleaned out
+
+        # rareSeekingIntentUserMap --> intent, user list
+
+        # comparing
+        for (
+            intent_code,
+            giver_user_list,
+        ) in (
+            rareGivingIntentUserMap.items()
+        ):  # I think something is failing here line 89/90
+            seeked_user_list = rareSeekingIntentUserMap[
+                intent_code
+            ]  # maybe not appending properly but I can't seem to find the issue
+
+            for seeker_user in seeked_user_list:
+
+                for giver_user in giver_user_list:
+                    rare_match = Match(
+                        users={
+                            seeker_user.uid,
+                            giver_user.uid,
+                        },
+                        metadata=MatchMetadata(
+                            generator="rareIntentsGenerator",
+                            rareIntents=[
+                                IntentMatch(
+                                    code=intent_code,
+                                    seeker=seeker_user.uid,
+                                    giver=giver_user.uid,
+                                )
+                            ],
+                        ),
+                    )
+                    user_one_in_match = seeker_user.uid
+                    user_two_in_match = giver_user.uid
+
+                    checking_same_user = False
+
+                    if int(user_one_in_match) > int(user_two_in_match):
```

Comment:
helpful to push this into a separate function and also leave a comment about what this if-else snippet is trying to achieve. Could also do this in a single line using the sort function - `''.join([user_one_in_match, user_two_in_match].sort())`